subsys/llext/fs_loader.c

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2024 BayLibre SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include <zephyr/llext/fs_loader.h>
+#include <zephyr/sys/util.h>
+#include <string.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(llext_fs_loader, CONFIG_LLEXT_LOG_LEVEL);
+
+int llext_fs_prepare(struct llext_loader *l)
+{
+	int ret = 0;
+	struct llext_fs_loader *fs_l = CONTAINER_OF(l, struct llext_fs_loader, loader);
+
+	fs_file_t_init(&fs_l->file);
+
+	ret = fs_open(&fs_l->file, fs_l->name, FS_O_READ);
+	if (ret != 0) {
+		LOG_DBG("Failed opening a file: %d", ret);
+		return ret;
+	}
+
+	fs_l->is_open = true;
+	return 0;
+}
+
+int llext_fs_read(struct llext_loader *l, void *buf, size_t len)
+{
+	int ret = 0;
+	struct llext_fs_loader *fs_l = CONTAINER_OF(l, struct llext_fs_loader, loader);
+
+	if (fs_l->is_open) {
+		ret = fs_read(&fs_l->file, buf, len);
+	} else {
+		ret = -EINVAL;
+	}
+
+	return ret == len ? 0 : ret;
```

Comment:
can `ret` be a positive number smaller than `len` in case of a partial read? If it can, then we probably also should return that as an error, because that's how `struct llext_loader::read` is defined